diff -ur apache-log4j-1.2.16.orig/build.xml apache-log4j-1.2.16/build.xml
--- apache-log4j-1.2.16.orig/build.xml	2010-03-31 00:14:39.000000000 -0400
+++ apache-log4j-1.2.16/build.xml	2020-09-14 14:32:04.417121378 -0400
@@ -41,7 +41,7 @@
   <!-- prefixed with "env". -->
   <property environment="env"/>
 
-  <property name="version" value="1.2.15"/>
+  <property name="version" value="1.2.16"/>
 
   <!-- The base directory relative to which most targets are built -->
   <property name="base" value="."/>
@@ -99,8 +99,8 @@
   <!-- Directory for temporary files. -->
   <property name="dist.tmp" value="${dist.dir}/tmp"/>
   
-  <property name="javac.source" value="1.2"/>
-  <property name="javac.target" value="1.1"/>
+  <property name="javac.source" value="1.7"/>
+  <property name="javac.target" value="1.7"/>
   
   <!--  destination for generated documentation on Apache web host  -->
   <property name="apache.javadoc_dest" value="/www/logging.apache.org/log4j/docs"/>
@@ -276,7 +276,6 @@
                <fileset file="${jaxp.jaxp.jar}"/>
            </classpath>
     </javac>
-    <rmic base="${javac.dest}" classname="examples.NumberCruncherServer"/>
 
     <copy todir="${javac.dest}">
       <fileset dir="." includes="examples/lf5/**/*.properties"/>
diff -ur apache-log4j-1.2.16.orig/src/main/java/org/apache/log4j/helpers/SyslogWriter.java apache-log4j-1.2.16/src/main/java/org/apache/log4j/helpers/SyslogWriter.java
--- apache-log4j-1.2.16.orig/src/main/java/org/apache/log4j/helpers/SyslogWriter.java	2010-03-31 00:14:26.000000000 -0400
+++ apache-log4j-1.2.16/src/main/java/org/apache/log4j/helpers/SyslogWriter.java	2020-09-14 15:07:03.646904872 -0400
@@ -124,9 +124,9 @@
         //  syslog packets must be less than 1024 bytes
         //
         int bytesLength = bytes.length;
-        if (bytesLength >= 1024) {
-            bytesLength = 1024;
-        }
+        // if (bytesLength >= 1024) {
+        //     bytesLength = 1024;
+        // }
         DatagramPacket packet = new DatagramPacket(bytes, bytesLength,
                                address, port);
         ds.send(packet);
diff -ur apache-log4j-1.2.16.orig/src/main/java/org/apache/log4j/net/SyslogAppender.java apache-log4j-1.2.16/src/main/java/org/apache/log4j/net/SyslogAppender.java
--- apache-log4j-1.2.16.orig/src/main/java/org/apache/log4j/net/SyslogAppender.java	2010-03-31 00:14:31.000000000 -0400
+++ apache-log4j-1.2.16/src/main/java/org/apache/log4j/net/SyslogAppender.java	2020-09-14 13:28:40.352008508 -0400
@@ -332,11 +332,12 @@
     //
     //   if message has a remote likelihood of exceeding 1024 bytes
     //      when encoded, consider splitting message into multiple packets
-    if (packet.length() > 256) {
-        splitPacket(hdr, packet);
-    } else {
-        sqw.write(packet);
-    }
+    // if (packet.length() > 256) {
+    //     splitPacket(hdr, packet);
+    // } else {
+    //     sqw.write(packet);
+    // }
+    sqw.write(packet);
 
     if (layout == null || layout.ignoresThrowable()) {
       String[] s = event.getThrowableStrRep();
