diff -Naur python-jsonrpc-r19/jsonrpc/json.py python-jsonrpc-r19.patched/jsonrpc/json.py
--- python-jsonrpc-r19/jsonrpc/json.py  2008-09-12 13:28:31.000000000 -0700
+++ python-jsonrpc-r19.patched/jsonrpc/json.py  2013-05-29 13:00:00.000000000 -0700
@@ -77,7 +77,7 @@
             return c
 
 def dumps(obj):
-    return unicode("".join([part for part in dumpParts (obj)]))
+    return "".join([part for part in dumpParts (obj)])
 
 def dumpParts (obj):
     objType = type(obj)
@@ -96,13 +96,22 @@
                 isFirst=False
             else:
                 yield u","
-            yield u'"' + StringEscapeRE.sub(escapeChar, key) +u'":'
+            decoded = None
+            try:
+                decoded = unicode(key,'utf-8')
+            except (UnicodeDecodeError,UnicodeEncodeError,TypeError):
+               decoded = unicode(key)
+            yield u'"' + StringEscapeRE.sub(escapeChar, decoded) +u'":'
             for part in dumpParts (value):
                 yield part
         yield u'}'
     elif objType in StringTypes:
-        yield u'"' + StringEscapeRE.sub(escapeChar, obj) +u'"'
-
+	decoded =None
+	try:
+	    decoded=unicode(obj,'utf-8')
+        except (UnicodeDecodeError,UnicodeEncodeError,TypeError):
+            decoded = unicode(obj)
+        yield  u'"' + StringEscapeRE.sub(escapeChar, decoded) +u'"'
     elif objType in [TupleType, ListType, GeneratorType]:
         yield u'['
         isFirst=True
@@ -121,8 +130,13 @@
     
 
 def loads(s):
+    decoded = None
+    try:
+        decoded = unicode(s,'utf-8')
+    except (TypeError,UnicodeDecodeError,UnicodeEncodeError):
+        decoded = s
     stack = []
-    chars = iter(s)
+    chars = iter(decoded)
     value = None
     currCharIsNext=False
 
@@ -228,3 +242,4 @@
          raise JSONDecodeException("Unexpected end of JSON source")
 
 
+
