diff -urN --ignore-all-space linux-2.6.12.2-orig/include/linux/netfilter_ipv4/ipt_physdev.h linux-2.6.12.2-working/include/linux/netfilter_ipv4/ipt_physdev.h
--- linux-2.6.12.2-orig/include/linux/netfilter_ipv4/ipt_physdev.h	2005-06-29 16:00:53.000000000 -0700
+++ linux-2.6.12.2-working/include/linux/netfilter_ipv4/ipt_physdev.h	2005-07-05 15:08:34.450593616 -0700
@@ -13,12 +13,12 @@
 #define IPT_PHYSDEV_OP_MASK		(0x20 - 1)
 
 struct ipt_physdev_info {
+	u_int8_t invert;
+	u_int8_t bitmask;
 	char physindev[IFNAMSIZ];
 	char in_mask[IFNAMSIZ];
 	char physoutdev[IFNAMSIZ];
 	char out_mask[IFNAMSIZ];
-	u_int8_t invert;
-	u_int8_t bitmask;
 };
 
 #endif /*_IPT_PHYSDEV_H*/
